### PR TITLE
ocamlformat-rpc: improve the documentation of the API

### PR DESCRIPTION
**Pull request by gpetiot**
Follow up on https://github.com/ocaml-ppx/ocamlformat/pull/1975#issuecomment-1025358582

- README for `ocamlformat-rpc`, detailing the commands (the commands handling csexp)
- README for `ocamlformat-rpc-lib`, explaining the way the API is supposed to be used (not using csexp)
- doc comments in `ocamlformat_rpc_lib.mli`

### CODE REVIEW

File: lib-rpc/README.md [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+  val return : 'a -> 'a t
+  (** [return a] will construct a constant IO value. *)
+
+  val read : ic -> Csexp.t option t
+  val write : oc -> Csexp.t list -> unit t
+```
+
+A basic implementation of this module can be:
+```ocaml
+module IO = struct
```
Comment (Julow) [inline lib-rpc/README.md:23]: A flaw with the library is that the module `IO` isn't accessible. Otherwise, this would be better to avoid weird error messages:
```ocaml
module IO : Ocamlformat_rpc_lib.IO = struct
```

File: lib-rpc/README.md [hunk 2]
```diff
@@ -0,0 +1,86 @@
+# ocamlformat-rpc-lib
```
Comment (Julow) [inline lib-rpc/README.md:1]: All this file would be better written in ocamlformat_rpc_lib.mli to take advantage of Odoc and avoid repeating code.

**Review by Julow — COMMENTED** [review]
I'm not fan of writing all this in README files. I think the doc for `lib-rpc` would be better in the `.mli` and the doc for `ocamlformat-rpc` would be better as a `.mld` page.

Using Odoc has many advantages:
- References
- A single way to distribute documentation. Odoc doesn't read the README files at all so there's two separate documentation packages: Odoc and browsing Github. Only Odoc documentation can be deployed with dune-release.
- Work with v3.ocaml.org and odig. Both tools read the root readme but not nested ones.

File: lib-rpc/README.md (hunk 1 above)
Comment (Julow) [reply lib-rpc/README.md:23]: Well, not really because we need to make `ic` and `oc` escape. That would look too complex for an example.

File: lib-rpc/README.md (hunk 2 above)
Comment (gpetiot) [reply lib-rpc/README.md]: done

**Comment by gpetiot** [comment]
I moved the `ocamlformat-rpc` doc to a .mld file, I wasn't sure whether to put it in doc/ or where the binary is.

File: bin/ocamlformat-rpc/README.md [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+All versions support the following commands:
+- `Halt` to end the communication with the RPC server. The caller must close the input and output channels.
+
+Some RPC versions offer specific commands, that are detailed below.
+
+### Version 1
+
+Specific commands supported on version v1 are:
+- `Config <csexp>`: submits a list of (key, value) pairs (as a canonical s-expression) to update OCamlFormat's configuration (please refer to `ocamlformat --help` to know more about the available options). The accepted configuration is sent as a reply of the same form. The configuration can be reset to its default value by sending the pair `("profile", "default").
+- `Format <csexp>`: submits a fragment of OCaml code (as a canonical s-expression) to be formatted by OCamlFormat, the formatted output is sent as a reply of the same form `Format <csexp>`"
```
[the remark this answers is not in the document]
Comment (gpetiot) [reply bin/ocamlformat-rpc/README.md]: done

File: bin/ocamlformat-rpc/index.mld [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+- [Format <s>]: submits a fragment of OCaml code [s] to be formatted by OCamlFormat, [ocamlformat-rpc] will attempt to seccessively format [s] as a [core_type], a [signature], a [module_type], an [expression] and a [use_file], stopping at the first successful formatting. The command expects a canonical s-expression of the form [List [Atom "Format"; Atom x]]. The formatted output is sent as a reply of the same form.
+
+{2 Version 2}
+
+Specific commands supported on version v2 are:
+- [Format <s>]: submits a canonical s-expression encoding a list where the first element of the list is a string to be formatted by OCamlFormat (similarly to the [Format] command of v1). The other arguments are (key, value) pairs, where key can be either ["Path"] and/or ["Config"]. They modify the server's configuration temporarily, for the current request. The formatted output is sent as a reply of the same form.
```
Comment (Julow) [inline bin/ocamlformat-rpc/index.mld:33]: This part is a bit hard to follow. Perhaps written as a pseudo-example would be better.

File: bin/ocamlformat-rpc/index.mld [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+
+Unknown commands are ignored.
+
+To build your RPC client to interact with [ocamlformat-rpc] please refer to [ocamlformat-rpc-lib]'s {{:../../lib-rpc/ocamlformat_rpc_lib.mli}documentation}.
```
Comment (Julow) [inline bin/ocamlformat-rpc/index.mld:37]: I would like to write this:

```
To build your RPC client to interact with [ocamlformat-rpc] please refer to {!Ocamlformat_rpc_lib}.
```

but it seems to not be working. Anyway, this link won't ever be valid.

```suggestion
To build your RPC client to interact with [ocamlformat-rpc] please refer to the library [ocamlformat-rpc-lib].
```

**Review by Julow — COMMENTED** [review]
I have made some improvements to the doc here: https://github.com/gpetiot/ocamlformat/pull/9

File: bin/ocamlformat-rpc/index.mld [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+
+All versions support the following commands:
+- [Halt] to end the communication with the RPC server. The caller must close the input and output channels.
+
+Some RPC versions offer specific commands, that are detailed below.
+
+{2 Version 1}
+
+Specific commands supported on version v1 are:
+- [Config <key, value>]: submits a list of (key, value) pairs to update OCamlFormat's configuration (please refer to [ocamlformat --help] to know more about the available options). The command expects a canonical s-expression of the form [List [Atom "Config"; List l]] where [l] is a list of canonical s-expressions encoding the key-values. The accepted configuration is sent as a reply of the same form. The configuration can be reset to its default value by sending the pair [("profile", "default")].
```
Comment (Julow) [inline bin/ocamlformat-rpc/index.mld:27]: Let's stop saying "canonical s-expression" everywhere. We need a paragraph saying that all the commands take canonical s-expressions.

```suggestion
- [(Config (options...))]: submits a list of formatting options (please refer to [ocamlformat --help] to know more about the available options). The server answer with the accepted configuration. The configuration can be reset to its default value by sending the pair [(profile default)].
```

File: bin/ocamlformat-rpc/index.mld [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+- [Format <s>]: submits a string to be formatted by OCamlFormat. The command expects a canonical s-expression of the form [List [Atom "Format"; Atom x]]. [ocamlformat-rpc] will attempt to successively format [x] as a [core_type], a [signature], a [module_type], an [expression] and a [use_file], stopping at the first successful formatting. The formatted output is sent as a reply of the same form.
```
Comment (Julow) [inline bin/ocamlformat-rpc/index.mld:28]:
```suggestion
- [(Format x)]: submits a string to be formatted. It attempts to successively format [x] as a [core_type], a [signature], a [module_type], an [expression] and a [use_file], stopping at the first successful formatting. The formatted output is sent as a reply of the same form.
```

File: bin/ocamlformat-rpc/index.mld [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+- [Version <v>] to ask the server to use version [v]. The command expects a canonical s-expression of the form [List [Atom "Version"; Atom v]] where [v] is of the version name (e.g. ["v1"]).
+
+If the server agrees upon the version he will send the reply [Version v] and the protocol version is set to [v], to use another version later the client has to close the connexion and start a new one. If the server cannot use version [v] he might propose another version [w] by sending the reply [Version w] that the client can accept by sending the same request for version [w], or propose another version. If the server cannot propose another version it will close the connection. Unknown commands are ignored.
+
+Once the client and the server agree on a common version, the requests you can send may differ from one version to another.
+
+{1 Commands}
+
+All versions support the following commands:
+- [Halt] to end the communication with the RPC server. The caller must close the input and output channels.
```
Comment (Julow) [inline bin/ocamlformat-rpc/index.mld:20]: I think this section isn't useful, it could be copied into both sections below. Also, something tells me this command is useless if it waits for the input to be closed.

File: bin/ocamlformat-rpc/index.mld [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+
+{2 Version 2}
+
+Specific commands supported on version v2 are:
+- [Format <s>]: submits a canonical s-expression encoding the string [x] to be formatted by OCamlFormat (similarly to the [Format] command of v1) and optional arguments modify the server's configuration temporarily, for the current request. Here are the different allowed forms:
+  + [List (Atom "Format"; Atom x)]
+  + [List (Atom "Format"; Atom x;  List (Atom "Path"; Atom p))]
+  + [List (Atom "Format"; Atom x; List (Atom "Config"; List l))]
+  + [List (Atom "Format"; Atom x; List (Atom "Path"; Atom p); List (Atom "Config"; List l))]
+ The [Config] argument is treated the same as the [Config] command of v1, except the modification only affects the current request. The formatted output is sent as a reply of the same form.
```
Comment (Julow) [inline bin/ocamlformat-rpc/index.mld:38]: I think it's a bit too formal. Also, no need to say "temporary":
```suggestion
- [(Format x options...)]: submits a string to be formatted. It attempts to successively format [x] as a [core_type], a [signature], a [module_type], an [expression] and a [use_file], stopping at the first successful formatting. The formatted output is sent as a reply of the same form.
  The options are:
  + [(Path p)] sets the path to the formatted file.
  + [(Config config)] sets the formatting options. [config] is a list of options, for example:
    [((profile default) (parse-docstrings true))]
    See [ocamlformat --help] for the available formatting options.
```
Move explanation for the `Path` option could be great.

File: bin/ocamlformat-rpc/index.mld (hunk 8 above)
Comment (gpetiot) [reply bin/ocamlformat-rpc/index.mld]: done
